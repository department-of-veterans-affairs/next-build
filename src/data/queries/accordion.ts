import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAccordion } from '@/types/drupal/paragraph'
import { Accordion } from '@/types/formatted/accordion'

export const formatter: QueryFormatter<ParagraphAccordion, Accordion> = (
  entity: ParagraphAccordion
) => {
  return {
    id: entity.id,
    header: entity.field_header || null,
    html: entity.field_rich_wysiwyg.processed,
  }
}
