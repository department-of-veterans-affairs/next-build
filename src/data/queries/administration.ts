import { FieldAdministration } from '@/types/drupal/field_type'

export interface Administration {
  entityId: number
  name: string | null
}

/**
 * Formats a Drupal administration field into our standardized Administration interface.
 * 
 * In our Drupal content, the `field_administration` field points to a taxonomy Term
 * that is currently called "Section", but "Section" is only a human label added to
 * it in the Drupal interface, and that name has evolved over time. It was originally,
 * "administration" (and thus the name of the field). Since the only machine name of
 * it is `field_administration`, we'll continue to call it "administration" in our code.
 */
export const formatAdministration = (
  administration: FieldAdministration | null | undefined
): Administration => ({
  entityId: administration?.drupal_internal__tid || null, // "t" is for taxonomy term
  name: administration?.name || null,
})
