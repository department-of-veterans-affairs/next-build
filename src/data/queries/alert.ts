// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { Alert } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, Alert> = (
  entity: BlockAlert
) => {
  return {
    alertType:
      entity.field_alert_type === 'information'
        ? 'info'
        : entity.field_alert_type,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header: entity.field_alert_content.field_text_expander,
      text: entity.field_alert_content.field_wysiwyg.processed,
    },
  }
}
