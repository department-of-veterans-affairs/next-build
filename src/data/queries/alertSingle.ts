import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAlertSingle } from '@/types/drupal/paragraph'
import {
  AlertSingle,
  AlertBlock,
  AlertNonReusable,
} from '@/types/formatted/alert'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '.'

// TODO:
//
// EDIT:
// It seems *maybe* that next-drupal-query is designed to handle this already.
// You can pass an `id` to queries.getParams().
// E.g.
// queries.getParams('paragraph--alert_single'))
//
// We don't need this here:
// export const params: QueryParams<null> = () => {
//   return queries
//     .getParams()
//     .addInclude([
//       'field_alert_block_reference',
//       'field_alert_block_reference.field_alert_content',
//       'field_alert_non_reusable_ref',
//       'field_alert_non_reusable_ref.field_va_paragraphs',
//     ])
// }
//
// ...but it seems like we might want to define
// paragraph-specific `includes` within the paragraph's
// query file (this file), and then somehow reference these in
// the node query when we need to include fields of a specific
// paragraph type.
//
// E.g.
// export const include = [
//   'field_alert_block_reference',
//   'field_alert_block_reference.field_alert_content',
//   'field_alert_non_reusable_ref',
//   'field_alert_non_reusable_ref.field_va_paragraphs',
// ]
//
// /* Some util file*/
// export const getIncludedSubFields = (
//   parentField: string,
//   subFields: string[]
// ) => [parentField, ...subFields.map((subField) => `${parentField}.${subField}`)]
//
// /* src/data/queries/resourcesSupport.ts */
// import { include } from '@/data/queries/alertSingle'
// export const params: QueryParams<null> = () => {
//   return queries.getParams().addInclude([
//     ...getIncludedSubFields('field_alert_single', include)
//   ])
// }

export const formatter: QueryFormatter<ParagraphAlertSingle, AlertSingle> = (
  entity: ParagraphAlertSingle
) => {
  return {
    type: entity.type as AlertSingle['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertSelection:
      entity.field_alert_selection as AlertSingle['alertSelection'],
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    nonReusableRef: formatParagraph(
      entity.field_alert_non_reusable_ref
    ) as AlertNonReusable,
  }
}
