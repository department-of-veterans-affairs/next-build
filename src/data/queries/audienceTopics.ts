// Define the query params for fetching node--news_story.
import { ParagraphAudienceTopics } from '@/types/drupal/paragraph'
import { QueryFormatter } from 'next-drupal-query'
import { AudienceTopic, AudienceTopics } from '@/types/formatted/audienceTopics'

const getTagsList = (
  entity: ParagraphAudienceTopics
): AudienceTopic[] | null => {
  if (!entity) return null

  const {
    field_topics: fieldTopics = [],
    field_audience_beneficiares: fieldAudienceBeneficiares,
    field_non_beneficiares: fieldNonBeneficiares,
  } = entity

  const topics = fieldTopics.map((topic) => ({
    id: topic.id,
    href: topic.path?.alias,
    name: topic.name,
    categoryLabel: 'Topics',
  }))

  const audiences = [
    ...(fieldAudienceBeneficiares || []),
    ...(fieldNonBeneficiares || []),
  ]
    .filter((tag) => !!tag)
    .map((audience) => ({
      id: audience.id,
      href: audience.path?.alias,
      name: audience.name,
      categoryLabel: 'Audience',
    }))

  const tagList = [...topics, ...audiences]

  return tagList.sort((a, b) => a.categoryLabel.localeCompare(b.categoryLabel))
}

/**
 * This formatter is unusual in that it takes a paragraph entity and digs into
 * its data to return multiple instances of similar values from it.
 */
export const formatter: QueryFormatter<
  ParagraphAudienceTopics,
  AudienceTopics
> = (entity: ParagraphAudienceTopics) => {
  return {
    type: entity.type as AudienceTopics['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    tags: getTagsList(entity),
  }
}
