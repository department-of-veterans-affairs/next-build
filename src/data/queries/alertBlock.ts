// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertBlock, AlertType } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, AlertBlock> = (
  entity: BlockAlert
) => {
  if (!entity) {
    return null
  }

  return {
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header:
        entity.field_alert_content.field_text_expander ??
        entity.field_alert_heading,
      text: entity.field_alert_content?.field_wysiwyg?.processed ?? '',
    },
  }
}
