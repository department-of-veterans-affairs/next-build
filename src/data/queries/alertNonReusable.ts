// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertType, AlertNonReusable } from '@/types/formatted/alert'
import { ParagraphNonReusableAlert } from '@/types/drupal/paragraph'
import { formatParagraph } from '@/lib/drupal/paragraphs'

export const formatter: QueryFormatter<
  ParagraphNonReusableAlert,
  AlertNonReusable
> = (entity: ParagraphNonReusableAlert) => {
  if (!entity) {
    return null
  }

  return {
    type: entity.type as AlertNonReusable['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    heading: entity.field_alert_heading,
    paragraphs: entity.field_va_paragraphs.map(formatParagraph),
  }
}
