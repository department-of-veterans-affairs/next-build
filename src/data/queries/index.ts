import { createQueries } from 'next-drupal-query'
import * as NewsStory from './newsStory'
import * as NewsStoryTeaser from './newsStoryTeaser'
import * as StoryListing from './storyListing'
import * as QuestionAnswer from './questionAnswer'
import * as ExpandableText from './expandableText'
import * as LinkTeaser from './linkTeaser'
import * as MediaImage from './mediaImage'
import * as Banner from './banners'
import * as PersonProfile from './personProfile'
import * as Button from './button'
import * as AudienceTopics from './audienceTopics'
import * as Alert from './alert'
import * as EmailContact from './emailContact'
import * as BenefitsHub from './benefitsHub'
import * as Wysiwyg from './wysiwyg'
import * as StaticPathResources from './staticPathResources'
import * as HeaderFooter from './headerFooter'

export const queries = createQueries({
  // standard Drupal entity data queries
  'node--news_story': NewsStory,
  'node--news_story--teaser': NewsStoryTeaser,
  'node--story_listing': StoryListing,
  'node--q_a': QuestionAnswer,
  'node--person_profile': PersonProfile,
  'node--landing_page': BenefitsHub, // "Benefits Hub Landing Page"
  'paragraph--audience_topics': AudienceTopics,
  'paragraph--button': Button,
  'paragraph--email_contact': EmailContact,
  'paragraph--expandable_text': ExpandableText,
  'paragraph--link_teaser': LinkTeaser,
  'paragraph--rich_text_char_limit_1000': Wysiwyg,
  'paragraph--wysiwyg': Wysiwyg,
  'media--image': MediaImage,
  'block--alert': Alert,

  // Custom queries 
  'banner--alerts_lookup': Banner,
  'header-footer-data': HeaderFooter,

  // Static Path Generation
  'static-path-resources': StaticPathResources,
})
