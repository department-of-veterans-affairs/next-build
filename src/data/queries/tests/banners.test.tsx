import { NodeBanner } from '@/types/drupal/node'
import { queries } from '@/data/queries/'
import mockData from '@/mocks/banners.mock.json'

//eslint-disable-next-line
const nodeBannerMock: NodeBanner | any = mockData

describe('Banners return formatted data', () => {
  let windowSpy

  beforeEach(() => {
    windowSpy = jest.spyOn(window, 'window', 'get')
  })

  afterEach(() => {
    windowSpy.mockRestore()
  })

  test('outputs formatted data', () => {
    windowSpy.mockImplementation(() => undefined)
    const formattedData = nodeBannerMock
    expect(
      queries.formatData('banner--alerts_lookup', formattedData)
    ).toMatchSnapshot()
  })
})
