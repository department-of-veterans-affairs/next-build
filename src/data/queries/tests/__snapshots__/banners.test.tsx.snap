// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banners return formatted data outputs formatted data 1`] = `
Array [
  Object {
    "alertType": "warning",
    "body": Object {
      "format": "rich_text_limited",
      "processed": "<p><strong>Natural disaster affecting your area:</strong> Use the <a href=\\"https://www.va.gov/find-locations\\">Find VA locations tool</a> to check the status and operating hours of your local VA facility.</p>
<p><strong>Coronavirus information:</strong> To get the latest updates and sign up to stay informed about COVID-19 vaccines, visit our <a href=\\"https://www.va.gov/health-care/covid-19-vaccine/\\">vaccine information page</a>. </p>
",
      "summary": "",
      "value": "<p><strong>Natural disaster affecting your area:</strong>&nbsp;Use the&nbsp;<a href=\\"https://www.va.gov/find-locations\\">Find VA locations tool</a>&nbsp;to check the status and operating hours of your local VA facility.</p>

<p><strong>Coronavirus information:</strong>&nbsp;To get the latest updates and sign up to stay informed about COVID-19 vaccines, visit our&nbsp;<a href=\\"https://www.va.gov/health-care/covid-19-vaccine/\\">vaccine information page</a>.&nbsp;</p>
",
    },
    "dismiss": true,
    "id": "6a273106-acfa-4615-91d4-6209d5af6bc4",
    "title": "VA facility and status updates",
    "type": "node--banner",
  },
  Object {
    "alertType": "announcement",
    "href": "entity:node/47278",
    "id": "ca11d747-65c8-4c10-adf2-2117726b0afb",
    "title": "Learn what the PACT Act means for your VA benefits",
    "type": "node--promo_banner",
  },
  Object {
    "alertType": "information",
    "bannerAlertVacms": Array [
      Object {
        "id": "8822bbdd-f5f8-4700-8abb-5566431bd705",
        "resourceIdObjMeta": Object {
          "drupal_internal__target_id": 1010,
        },
        "type": "node--vamc_operating_status_and_alerts",
      },
    ],
    "body": Object {
      "format": "rich_text",
      "processed": "<p><strong>IMPORTANT UPDATES FROM VA PITTSBURGH</strong></p>

<p>In-person visitation for COVID-negative inpatients at University Drive is limited to two visitors per patient during specified hours. Outpatients can bring a support person to clinic appointments and procedures. We require all entrants to wear a VA-provided procedure mask. For more information, please see <a href=\\"https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.va.gov%2Fpittsburgh-health-care%2Foperating-status%2F&amp;data=04%7C01%7C%7C98a2764fa768480e295f08d9e5b0a0d9%7Ce95f1b23abaf45ee821db7ab251ab3bf%7C0%7C0%7C637793369034143792%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&amp;sdata=238BicloaB08fgpaagfUeVV24lrJoSQpAXEFZCaLbBs%3D&amp;reserved=0\\">Operating Status</a>. Contact us before you visit: For some needs, we may be able to provide care by phone or video.</p>

<p><strong>COVID-19 Vaccines at VA Pittsburgh:</strong> For information on vaccine eligibility and availability at our medical campuses and outpatient clinics, please see <a href=\\"https://www.va.gov/pittsburgh-health-care/covid-19-vaccines/\\">VAPHS COVID-19 Vaccines</a>.</p>

<p><strong>COVID-19 Emergency Department care:</strong> To ensure we can prioritize treatment for Veterans most in need of emergency COVID-19 care, do not report to our Emergency Department for COVID-19 treatment unless you have trouble breathing or severe symptoms. Do not report to our Emergency Department for COVID-19 treatment/testing if you have mild or no symptoms. <strong>Do not walk-in for non-emergency care:</strong> To schedule an appointment, call 412-822-3000 or send a secure message to your provider via MyHealtheVet at myhealth.va.gov.</p>

<p><strong>COVID-19 testing:</strong> COVID-19 testing is by appointment only. If you have a sore throat, dry cough, shortness of breath or fever, call 412-360-6000 and select option #3 to speak to a nurse before visiting our facilities or to schedule an appointment for testing at University Drive.</p>

<p><strong>For the latest coronavirus information: </strong>Visit <a href=\\"https://www.cdc.gov/coronavirus/2019-ncov/\\">CDC: What you Need to Know</a></p>

<p><strong>For VA-specific information:</strong> Visit <a href=\\"https://www.va.gov/health-care/covid-19-vaccine/\\">COVID-19 Vaccines at VA</a>, use VA’s <a href=\\"https://www.va.gov/coronavirus-chatbot/\\">Coronavirus Chatbot</a> or read <a href=\\"https://www.va.gov/coronavirus-veteran-frequently-asked-questions/\\">Coronavirus FAQs</a> and <a href=\\"https://www.publichealth.va.gov/n-coronavirus/\\">Novel Coronavirus Disease (COVID-19)</a>.         </p>",
      "value": "<p><strong>IMPORTANT UPDATES FROM VA PITTSBURGH</strong></p>

<p>In-person visitation for COVID-negative inpatients at University Drive is limited to two visitors per patient during specified hours. Outpatients can bring a support person to clinic appointments and procedures. We require all entrants to wear a VA-provided procedure mask.&nbsp;For more information, please see&nbsp;<a href=\\"https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.va.gov%2Fpittsburgh-health-care%2Foperating-status%2F&amp;data=04%7C01%7C%7C98a2764fa768480e295f08d9e5b0a0d9%7Ce95f1b23abaf45ee821db7ab251ab3bf%7C0%7C0%7C637793369034143792%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&amp;sdata=238BicloaB08fgpaagfUeVV24lrJoSQpAXEFZCaLbBs%3D&amp;reserved=0\\">Operating Status</a>. Contact us before you visit: For some needs, we may be able to provide care by phone or video.</p>

<p><strong>COVID-19 Vaccines at VA Pittsburgh:</strong> For information on vaccine eligibility and availability at our medical campuses and outpatient clinics, please see <a href=\\"https://www.va.gov/pittsburgh-health-care/covid-19-vaccines/\\">VAPHS COVID-19 Vaccines</a>.</p>

<p><strong>COVID-19 Emergency Department care:</strong> To ensure we can prioritize treatment for Veterans most in need of emergency COVID-19 care, do not report to our Emergency Department for COVID-19 treatment unless you have trouble breathing or severe symptoms. Do not report to our Emergency Department for COVID-19 treatment/testing if you have mild or no symptoms.&nbsp;<strong>Do not walk-in for non-emergency care:</strong> To schedule an appointment, call 412-822-3000 or send a secure message to your provider via MyHealtheVet at myhealth.va.gov.</p>

<p><strong>COVID-19 testing:</strong> COVID-19 testing is by appointment only. If you have a sore throat, dry cough, shortness of breath or fever, call 412-360-6000 and select option #3 to speak to a nurse before visiting our facilities or to schedule an appointment for testing at University Drive.</p>

<p><strong>For the latest coronavirus information: </strong>Visit <a href=\\"https://www.cdc.gov/coronavirus/2019-ncov/\\">CDC: What you Need to Know</a></p>

<p><strong>For VA-specific information:</strong> Visit <a href=\\"https://www.va.gov/health-care/covid-19-vaccine/\\">COVID-19 Vaccines at VA</a>, use VA’s <a href=\\"https://www.va.gov/coronavirus-chatbot/\\">Coronavirus Chatbot</a> or read <a href=\\"https://www.va.gov/coronavirus-veteran-frequently-asked-questions/\\">Coronavirus FAQs</a> and <a href=\\"https://www.publichealth.va.gov/n-coronavirus/\\">Novel Coronavirus Disease (COVID-19)</a>.&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</p>
",
    },
    "dismiss": true,
    "id": "ccd9d30f-78f9-4358-80d7-191f99b18d43",
    "inheritanceSubpages": false,
    "operatingStatus": true,
    "path": "/va-pittsburgh-health-care/vamc-banner-alert/2021-01-08/help-limit-the-spread-of-covid-19-and-other",
    "title": "Help Limit the Spread of COVID-19 and other Illnesses",
    "type": "node--full_width_banner_alert",
  },
  Object {
    "alertType": "announcement",
    "href": "entity:node/47278",
    "id": "ca11d747-65c8-4c10-adf2-2117726b0afb",
    "title": "Learn what the PACT Act means for your VA benefits",
    "type": "node--promo_banner",
  },
]
`;
